contracts/src/TeleporterTokenSource.sol

```diff
@@ -182,9 +317,54 @@ abstract contract TeleporterTokenSource is
         );
 
         if (isMultihop) {
-            emit TokensAndCallRouted(messageID, input, amount);
+            emit TokensAndCallRouted(messageID, input, adjustedAmount);
         } else {
-            emit TokensAndCallSent(messageID, originSenderAddress, input, amount);
+            emit TokensAndCallSent(messageID, originSenderAddress, input, adjustedAmount);
+        }
+    }
+
+    /**
+     * @dev See {INativeTokenSource-addCollateral}
+     */
+    function _addCollateral(
+        bytes32 destinationBlockchainID,
+        address destinationBridgeAddress,
+        uint256 amount
+    ) internal sendNonReentrant {
+        DestinationBridgeSettings memory destinationSettings =
+            registeredDestinations[destinationBlockchainID][destinationBridgeAddress];
+        require(
+            destinationSettings.registered,
+            "TeleporterTokenSource: destination bridge not registered"
+        );
+        require(
+            destinationSettings.collateralNeeded > 0,
+            "TeleporterTokenSource: zero collateral needed"
+        );
+
+        // Deposit the full amount, and withdraw back to the sender if there is excess.
+        amount = _deposit(amount);
+
+        // Calculate the amount remaining.
+        uint256 remainingAmount;
+        uint256 excessAmount;
+        if (amount >= destinationSettings.collateralNeeded) {
+            remainingAmount = 0;
+            excessAmount = amount - destinationSettings.collateralNeeded;
+        } else {
+            remainingAmount = destinationSettings.collateralNeeded - amount;
+        }
+
+        // Update the reserve imbalance remaining.
+        registeredDestinations[destinationBlockchainID][destinationBridgeAddress].collateralNeeded =
+            remainingAmount;
+        emit CollateralAdded(
```

Comment:
Should we emit the amount added as `amount - excessAmount` here, since the excess is not effectively "added" now that it's returned?

Reply:
updated

Reply:
and added withdrawal check to unit test